File: TextosJapones/src/main/java/org/jisho/textosJapones/model/entities/Revisar.java
package org.jisho.textosJapones.model.entities;

import javafx.scene.control.CheckBox;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class Revisar {

    private UUID id;
    private String vocabulario;
    private String formaBasica;
    private String leitura;
    private String portugues;
    private String ingles;
    private Boolean anime;
    private Boolean manga;
    private Boolean novel;

    final private CheckBox revisado = new CheckBox();

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public String getVocabulario() {
        return vocabulario;
    }

    public void setVocabulario(String vocabulario) {
        this.vocabulario = vocabulario;
    }

    public String getFormaBasica() {
        return formaBasica;
    }

    public void setFormaBasica(String formaBasica) {
        this.formaBasica = formaBasica;
    }

    public String getLeitura() {
        return leitura;
    }

    public void setLeitura(String leitura) {
        this.leitura = leitura;
    }

    public String getPortugues() {
        return portugues;
    }

    public void setPortugues(String portugues) {
        this.portugues = portugues;
    }

    public String getIngles() {
        return ingles;
    }

    public void setIngles(String ingles) {
        this.ingles = ingles;
    }

    public CheckBox getRevisado() {
        return revisado;
    }

    public Boolean isAnime() {
        return anime;
    }

    public void setAnime(Boolean anime) {
        this.anime = anime;
    }

    public Boolean isManga() {
        return manga;
    }

    public void setManga(Boolean manga) {
        this.manga = manga;
    }

    public Boolean isNovel() {
        return novel;
    }

    public void setNovel(Boolean novel) {
        this.novel = novel;
    }


    public Revisar() {
        this.id = null;
        this.vocabulario = "";
        this.formaBasica = "";
        this.leitura = "";
        this.portugues = "";
        this.ingles = "";
        this.anime = false;
        this.manga = false;
        this.novel = false;
        this.revisado.setSelected(false);
    }

    public Revisar(String vocabulario) {
        this.id = null;
        this.vocabulario = vocabulario;
        this.formaBasica = "";
        this.leitura = "";
        this.portugues = "";
        this.ingles = "";
        this.anime = false;
        this.manga = false;
        this.novel = false;
        this.revisado.setSelected(false);
    }

    public Revisar(String vocabulario, String formaBasica, String leitura) {
        this.id = null;
        this.vocabulario = vocabulario;
        this.leitura = leitura;
        this.formaBasica = formaBasica;
        this.portugues = "";
        this.ingles = "";
        this.anime = false;
        this.manga = false;
        this.novel = false;
        this.revisado.setSelected(false);
    }

    public Revisar(String vocabulario, Boolean revisado, Boolean isAnime, Boolean isManga, Boolean isNovel) {
        this.id = null;
        this.vocabulario = vocabulario;
        this.leitura = "";
        this.formaBasica = "";
        this.portugues = "";
        this.ingles = "";
        this.anime = isAnime;
        this.manga = isManga;
        this.novel = isNovel;
        this.revisado.setSelected(revisado);
    }

    public Revisar(String vocabulario, String formaBasica, String leitura, Boolean revisado, Boolean isAnime, Boolean isManga, Boolean isNovel) {
        this.id = null;
        this.vocabulario = vocabulario;
        this.leitura = leitura;
        this.formaBasica = formaBasica;
        this.portugues = "";
        this.ingles = "";
        this.anime = isAnime;
        this.manga = isManga;
        this.novel = isNovel;
        this.revisado.setSelected(revisado);
    }

    public Revisar(String vocabulario, String formaBasica, String leitura, String portugues, String ingles) {
        this.id = null;
        this.vocabulario = vocabulario;
        this.formaBasica = formaBasica;
        this.leitura = leitura;
        this.portugues = portugues;
        this.ingles = ingles;
        this.anime = false;
        this.manga = false;
        this.novel = false;
        this.revisado.setSelected(false);
    }

    public Revisar(UUID id, String vocabulario, String formaBasica, String leitura, String portugues, String ingles,
                   Boolean revisado, Boolean anime, Boolean manga, Boolean novel) {
        this.id = id;
        this.vocabulario = vocabulario;
        this.formaBasica = formaBasica;
        this.leitura = leitura;
        this.portugues = portugues;
        this.ingles = ingles;
        this.anime = anime;
        this.manga = manga;
        this.novel = novel;
        this.revisado.setSelected(revisado);
    }

    @Override
    public String toString() {
        return vocabulario + ", ";
    }

    public static Vocabulario toVocabulario(Revisar revisar) {
        return new Vocabulario(null, revisar.getVocabulario(), revisar.getFormaBasica(), revisar.getLeitura(),
                revisar.getIngles(), revisar.getPortugues());
    }

    public static List<Vocabulario> toVocabulario(List<Revisar> revisar) {
        List<Vocabulario> lista = new ArrayList<>();

        for (Revisar obj : revisar)
            lista.add(toVocabulario(obj));

        return lista;
    }
}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/model/entities/processatexto/Revisar.kt
package br.com.fenix.processatexto.model.entities.processatexto

import br.com.fenix.processatexto.model.entities.EntityBase
import jakarta.persistence.Column
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import javafx.beans.value.ChangeListener
import javafx.scene.control.CheckBox
import org.hibernate.annotations.JdbcTypeCode
import org.hibernate.type.SqlTypes
import java.util.*


@jakarta.persistence.Entity
data class Revisar(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @JdbcTypeCode(SqlTypes.VARCHAR)
    @Column(name = "ID", nullable = false)
    private var id: UUID? = null,
    @Column(length = 250, nullable = false, unique = true)
    var vocabulario: String = "",
    @Column(name = "forma_basica", length = 250, nullable = true)
    var formaBasica: String = "",
    @Column(length = 250, nullable = true)
    var leitura: String = "",
    @Column(name = "leitura_novel", length = 250, nullable = true)
    var leituraNovel: String = "",
    @Column(nullable = true)
    var portugues: String = "",
    @Column(nullable = true)
    var ingles: String = "",
    @Column
    var aparece: Int = 0,
    @Transient
    private var _isRevisado: Boolean = false,
    @Column
    var isAnime: Boolean = false,
    @Column
    var isManga: Boolean = false,
    @Column
    var isNovel: Boolean = false
) : EntityBase<UUID?, Revisar>() {

    @Transient
    val revisado: CheckBox = CheckBox()

    @Transient
    private lateinit var listener: ChangeListener<Boolean>

    @Column(name = "revisado")
    var isRevisado: Boolean = _isRevisado
        set(value) {
            revisado.isSelected = isRevisado
            field = value
        }

    init {
        revisado.isSelected = isRevisado
        listener = ChangeListener<Boolean> { _, _, newValue ->
            try {
                revisado.selectedProperty().removeListener(listener)
                isRevisado = newValue
            } finally {
                revisado.selectedProperty().addListener(listener)
            }
        }
        revisado.selectedProperty().addListener(listener)
    }

    constructor(vocabulario: String) : this(null, vocabulario)

    constructor(vocabulario: String, formaBasica: String, leitura: String, leituraNovel: String) : this(null, vocabulario, formaBasica, leitura, leituraNovel)

    constructor(vocabulario: String, revisado: Boolean, isAnime: Boolean, isManga: Boolean, isNovel: Boolean) : this(null, vocabulario) {
        this.isAnime = isAnime
        this.isManga = isManga
        this.isNovel = isNovel
        this.revisado.isSelected = revisado
    }

    constructor(vocabulario: String, formaBasica: String, leitura: String, leituraNovel: String, revisado: Boolean, isAnime: Boolean, isManga: Boolean, isNovel: Boolean) :
            this(null, vocabulario, formaBasica, leitura, leituraNovel, "", "", 0, revisado, isAnime, isManga, isNovel) {
        this.revisado.isSelected = revisado
    }

    constructor(vocabulario: String, formaBasica: String, leitura: String, leituraNovel: String, portugues: String, ingles: String) :
            this(null, vocabulario, formaBasica, leitura, leituraNovel, portugues, ingles) {
    }

    override fun getId(): UUID? = id

    override fun create(id: UUID?): Revisar = Revisar(id)

    fun setId(id: UUID?) {
        this.id = id
    }

    @Override
    override fun toString(): String {
        return "$vocabulario, "
    }

    companion object {
        fun toVocabulario(revisar: Revisar): Vocabulario {
            return Vocabulario(revisar.id, revisar.vocabulario, revisar.formaBasica, revisar.leitura, revisar.leituraNovel, revisar.ingles, revisar.portugues)
        }

        fun toVocabulario(revisar: List<Revisar>): MutableList<Vocabulario> {
            val lista = mutableListOf<Vocabulario>()
            for (obj in revisar)
                lista.add(toVocabulario(obj))
            return lista
        }
    }
}